Building a NET6/NET7 iOS project on agent M2 ARM 64 hangs/freezes

### Steps to Reproduce

1. Create an agent on M2 ARM 64 agent (3.214.0)
2. Build a .NET6/.NET7 iOS project
3. Notice the build MIGHT hang sometimes on Apple Clang process

We don't face the issue on X64 on prem agents or even hosted. 
There is no real consistency on when the build will hangs or not. It depends on the run.

We already tried removing Trimmer, which doesn't seem to have any effect. With or without, the behavior is the same. 

### Expected Behavior

Build should never hang

### Actual Behavior

Build hangs sometimes and never ends, until timeout

### Environment

- Xcode 14.2
- Visual Studio for mac 17.5.1
- This is the [.csproj](https://github.com/nventive/UnoApplicationTemplate/blob/dev/vica/make-usage-new-agents-net7.0/src/app/ApplicationTemplate.Mobile/ApplicationTemplate.Mobile.csproj) that we try to build 

AGENT CAPABILITIES : 

<html>
<body>
<!--StartFragment-->

Agent.Name | MACOS-2C83F31C-42D1-4BA5-9686-611EB3632BD4 |   |  
-- | -- | -- | --
  | Agent.Version | 3.214.0 |   |  
  | _ | ./externals/node16/bin/node |   |  
  | __CF_USER_TEXT_ENCODING | 0x1F5:0x0:0x52 |   |  
 
  | CP_HOME_DIR | /Users/administrator/agent/_work/_temp/.cocoapods |   |  
  | curl | /usr/bin/curl |   |  
  | dotnet | /usr/local/share/dotnet/dotnet |   |  
  | DOTNET_ROOT | /usr/local/share/dotnet |   |  
  | git | /usr/bin/git |   |  
  | HOME | /Users/administrator |   |  
  | InteractiveSession | False |   |  
  | java | /usr/bin/java |   |  
  | JDK | /usr/bin/javac |   |  
  | LANG | en_CA.UTF-8 |   |  
  | LOGNAME | administrator |   |  
  | make | /usr/bin/make |   |  
  | MSBuild | /Library/Frameworks/Mono.framework/Versions/Current/Commands/msbuild |   |  
  | NUGET_HTTP_CACHE_PATH | /Users/administrator/agent/_work/_temp/.nuget-http-cache |   |  
  | NUGET_PACKAGES | /Users/administrator/agent/_work/_temp/.nuget |   |  
  | PATH | /Users/administrator/.rbenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/share/dotnet:~/.dotnet/tools:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands |   |  
  | PWD | /Users/administrator/agent |   |  
  | python3 | /usr/bin/python3 |   |  
  | rake | /Users/administrator/.rbenv/shims/rake |   |  
  | ruby | /Users/administrator/.rbenv/shims/ruby |   |  
  | sh | /bin/sh |   |  
  | SHELL | /bin/zsh |   |  
  | SSH_AUTH_SOCK | /private/tmp/com.apple.launchd.MgBJHUlv5M/Listeners |   |  
  | TMPDIR | /var/folders/33/ph0v51hd30n2frx557550mnc0000gn/T/ |   |  
  | USER | administrator |   |  
  | VSTS_AGENT_SVC | 1 |   |  
  | Xamarin.iOS | /Applications/Visual Studio.app/Contents/MacOS/vstool |   |  
  | Xamarin.iOS_Version | 16.1.1 |   |  
  | XamarinBuildDownloadDir | /Users/administrator/agent/_work/_temp/.xbcache |   |  
  | xcode | /Applications/Xcode.app/Contents/Developer |   |  
  | Xcode_Version | 14.2 |   |  
  | XPC_FLAGS | 0x0 |   |  
  | XPC_SERVICE_NAME | 0

<!--EndFragment-->
</body>
</html>

### Build Logs

- Working build logs => 
[iOS BUILD - OK.txt](https://github.com/xamarin/xamarin-macios/files/10991738/iOS.BUILD.-.OK.txt)

- Hanging build logs => 
[iOS BUILD - HANGS.txt](https://github.com/xamarin/xamarin-macios/files/10991734/iOS.BUILD.-.HANGS.txt)

**MSBUILD BINLOG** (seem corrupted ...)

[build-net7.0-ios.zip](https://github.com/xamarin/xamarin-macios/files/10992008/build-net7.0-ios.zip)

### Example Project (If Possible)

https://github.com/nventive/UnoApplicationTemplate/blob/dev/vica/make-usage-new-agents-net7.0/src/app/ApplicationTemplate.Mobile/ApplicationTemplate.Mobile.csproj

Thank you for your quick answer.

> Both of the stack traces contain OutOfProcNode.Run so they seem to be waiting for some other MSBuild (?) process.

As you saw I found two msbuild processes ... could it be that they wait on each other, ending in an endless waiting loop. Any advice maybe to try confirm that or seek for other processes that would be waited by msbuild ?

Finally ended ... 

```
[ERROR] System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.InternalRead(Int32 numBytes)
   at System.IO.BinaryReader.ReadUInt16()
   at Microsoft.Diagnostics.NETCore.Client.IpcHeader.Parse(BinaryReader[ERROR] System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at System.IO.BinaryReader.InternalRead(Int32 numBytes)
   at System.IO.BinaryReader.ReadUInt16()
   at Microsoft.Diagnostics.NETCore.Client.IpcHeader.Parse(BinaryReader reader) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcHeader.cs:line 55
   at Microsoft.Diagnostics.NETCore.Client.IpcMessage.Parse(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcMessage.cs:line 117
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.Read(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 107
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.SendMessageGetContinuation(IpcEndpoint endpoint, I reader) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcHeader.cs:line 55
   at Microsoft.Diagnostics.NETCore.Client.IpcMessage.Parse(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcMessage.cs:line 117
   apcMessage message) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 44
   at Microsoft.Diagnostics.NETCore.Client.EventPipeSession.Start(IpcEndpoint endpoint, IEnumerable`1 providers, Boolean requestRundown, Int32 circularBufferMB) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/EventPipeSession.cs:line 34
   at Microsoft.Diagnostics.Tools.Stack.ReportCommandHandler.Report(CancellationToken ct, IConsole console, Int32 processId, String name, TimeSpan duration)t Microsoft.Diagnostics.NETCore.Client.IpcClient.Read(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 107
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.SendMessageGetContinuation(IpcEndpoint endpoint, I
pcMessage message) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 44
   at Microsoft.Diagnostics.NETCore.Client.EventPipeSession.Start(IpcEndpoint endpoint, IEnumerable`1 providers, Boolean requestRundown, Int32 circularBufferMB) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/EventPipeSession.cs:line 34
   at Microsoft.Diagnostics.Tools.Stack.ReportCommandHandler.Report(CancellationToken ct, IConsole console, Int32 processId, String name, TimeSpan duration)
xargs: dotnet: exited with status 255; aborting
```

--- a/msbuild/Xamarin.MacDev.Tasks/Tasks/AOTCompileTaskBase.cs
+++ b/msbuild/Xamarin.MacDev.Tasks/Tasks/AOTCompileTaskBase.cs
@@ -176,13 +176,13 @@ namespace Xamarin.MacDev.Tasks {
 			Directory.CreateDirectory (OutputDirectory);
 
 			var aotAssemblyFiles = new List<ITaskItem> ();
-			var processes = new Task<Execution> [assembliesToAOT.Length];
 
 			var environment = new Dictionary<string, string?> {
 				{ "MONO_PATH", Path.GetFullPath (InputDirectory) },
 			};
 
 			var globalAotArguments = AotArguments?.Select (v => v.ItemSpec).ToList ();
+			var listOfArguments = new List<(IList<string> Arguments, string Input)> ();
 			for (var i = 0; i < assembliesToAOT.Length; i++) {
 				var asm = assembliesToAOT [i];
 				var input = asm.GetMetadata ("Input");
@@ -212,16 +212,20 @@ namespace Xamarin.MacDev.Tasks {
 				arguments.AddRange (parsedProcessArguments);
 				arguments.Add (input);
 
-				processes [i] = ExecuteAsync (AOTCompilerPath, arguments, environment: environment, sdkDevPath: SdkDevPath, showErrorIfFailure: false /* we show our own error below */)
-					.ContinueWith ((v) => {
-						if (v.Result.ExitCode != 0)
-							Log.LogError (MSBStrings.E7118 /* Failed to AOT compile {0}, the AOT compiler exited with code {1} */, Path.GetFileName (input), v.Result.ExitCode);
-
-						return System.Threading.Tasks.Task.FromResult<Execution> (v.Result);
-					}).Unwrap ();
+				listOfArguments.Add (new (arguments, input));
 			}
 
-			System.Threading.Tasks.Task.WaitAll (processes);
+			Parallel.ForEach (listOfArguments, (arg) => {
+				ExecuteAsync (AOTCompilerPath, arg.Arguments, environment: environment, sdkDevPath: SdkDevPath, showErrorIfFailure: false /* we show our own error below */)
+					.ContinueWith ((v) => {
+						if (v.Result.ExitCode != 0)
+							Log.LogError (MSBStrings.E7118 /* Failed to AOT compile {0}, the AOT compiler exited with code {1} */, Path.GetFileName (arg.Input), v.Result.ExitCode);
+
+						return System.Threading.Tasks.Task.FromResult<Execution> (v.Result);
+					})
+					.Unwrap ()
+					.Wait ();
+			});
 
 			AssemblyFiles = aotAssemblyFiles.ToArray ();
 

